File: src/integration/spring/src/jvmMain/java/community/flock/wirespec/integration/spring/java/configuration/WirespecSerializationConfiguration.java
package community.flock.wirespec.integration.spring.java.configuration;

import community.flock.wirespec.integration.spring.java.web.WirespecResponseBodyAdvice;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;

/**
 * Entry point imported by {@code @EnableWirespecController}. The Wirespec Serialization
 * bean and the multipart WirespecJsonMapper bean are contributed by one of the two
 * version-conditional configurations, so the integration runs on both Spring Boot 3
 * (Jackson 2) and Spring Boot 4 (Jackson 3). Jackson 3 takes precedence when both are present.
 */
@Configuration
@Import({
        WirespecResponseBodyAdvice.class,
        WirespecWebMvcConfiguration.class,
        WirespecNativeConfiguration.class,
        WirespecJackson3Configuration.class,
        WirespecJackson2Configuration.class,
})
public class WirespecSerializationConfiguration {
}

File: src/integration/spring/src/jvmMain/kotlin/community/flock/wirespec/integration/spring/java/configuration/WirespecSerializationConfiguration.kt
package community.flock.wirespec.integration.spring.java.configuration

import com.fasterxml.jackson.databind.ObjectMapper
import community.flock.wirespec.integration.jackson.java.WirespecSerialization
import community.flock.wirespec.integration.spring.java.web.WirespecResponseBodyAdvice
import community.flock.wirespec.java.Wirespec.Serialization
import community.flock.wirespec.java.serde.DefaultParamSerialization
import community.flock.wirespec.java.serde.DefaultPathSerialization
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import
import org.springframework.http.MediaType
import java.lang.reflect.Type

@Configuration
@Import(WirespecResponseBodyAdvice::class, WirespecWebMvcConfiguration::class)
open class WirespecSerializationConfiguration {

    @Bean
    open fun wirespecSerializationMap(
        objectMapper: ObjectMapper,
    ): Map<MediaType, Serialization> = mapOf(
        MediaType.APPLICATION_JSON to WirespecSerialization(objectMapper),
        MediaType.TEXT_PLAIN to PlainTextDataSerialization(),
    )
}

class PlainTextDataSerialization :
    Serialization,
    DefaultParamSerialization,
    DefaultPathSerialization {
    override fun <T : Any?> serializeBody(t: T?, type: Type?): ByteArray {
        return when (t) {
            is ByteArray -> return t
            is String -> return t.toByteArray()
            else -> error("Unsupported type ${t?.let { it::class.simpleName }}")
        }
    }
    override fun <T : Any?> deserializeBody(raw: ByteArray?, type: Type?): T = raw as T
}
